timer/src/main/java/com/toggl/timer/start/ui/StartTimeEntryDialogFragment.kt

```diff
@@ -0,0 +1,206 @@
+    }
+
+    private val descriptionChangeListener: TextWatcher = object : TextWatcher {
+        override fun afterTextChanged(text: Editable) {
+            val action = StartTimeEntryAction.DescriptionEntered(text.toString())
+            store.dispatch(action)
+        }
+
+        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
+
+        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
+    }
+
+    private val bottomSheetCallback = BottomSheetCallback()
+
+    override fun onAttach(context: Context) {
+        (requireActivity().applicationContext as TimerComponentProvider)
+            .provideTimerComponent().inject(this)
+        super.onAttach(context)
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setStyle(DialogFragment.STYLE_NORMAL, R.style.BottomSheetDialog)
+    }
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? = inflater.inflate(R.layout.fragment_dialog_start_time_entry, container, false)
+
+    @ExperimentalCoroutinesApi
+    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+        return (super.onCreateDialog(savedInstanceState) as BottomSheetDialog).also { bottomSheetDialog ->
+            bottomSheetDialog.setOnShowListener { dialogInterface ->
+                store.state
+                    .mapNotNull { it.editableTimeEntry }
+                    .take(1)
+                    .onEach { editableTimeEntry ->
+                        dialogInterface.attachBottomView(bottomSheetDialog, R.layout.bottom_control_panel_layout, editableTimeEntry)
+                    }
+                    .launchIn(lifecycleScope)
+            }
+        }
+    }
+
+    @ExperimentalCoroutinesApi
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        bottomSheetCallback.addOnSlideAction(AlphaSlideAction(extended_options, false))
+        bottomSheetCallback.addOnStateChangedAction(object : OnStateChangedAction {
+            override fun onStateChanged(sheet: View, newState: Int) {
+                extended_options.isInvisible = newState == BottomSheetBehavior.STATE_COLLAPSED || newState == BottomSheetBehavior.STATE_HIDDEN
+            }
+        })
+
+        with(bottomSheetBehavior) {
+            addBottomSheetCallback(bottomSheetCallback)
+            skipCollapsed = false
+            peekHeight = resources.getDimension(R.dimen.time_entry_edit_half_expanded_height).toInt()
+            state = BottomSheetBehavior.STATE_COLLAPSED
+        }
+
+        with(time_entry_description) {
+            setOnFocusChangeListener { _, _ ->
+                post {
+                    (activity?.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager).apply {
```

Comment:
Would be nice to have an extension like we do in Xamarin to avoid such casts:

```kotlin
fun <T> Activity.getSystemService(service: String) : T =
    getSystemService(service) as T
```

So this would become:
```suggestion
                    activity?.getSystemService<InputMethodManager>(Context.INPUT_METHOD_SERVICE).apply {
```

Reply:
Also, why aren't you using `requireActivity`?

Reply:
Since this is happening inside of a listener and a post I would rather be safe and use safe access rather than to risk the possibility of NPE in the future just because of some non-essential keyboard logic.